[1단계 - 지하철 노선도 미션] 브랜(서지환) 미션 제출합니다.

seojihwan opened the pull request:
광휘님 안녕하세요~~ 브랜입니다!

마지막 미션인데 너무 바빠 뒤늦게 제출합니다ㅠㅠ 😭 

## 📣링크
[데모페이지](https://react-subwaymap.netlify.app/)
- 아이디: test@test.com 비밀번호: 123123

[스토리북](https://react-subwaymap-storybook.netlify.app/)
[VSCODE in web](http://github1s.com/seojihwan/react-subway-map/tree/step1)

### 로그인
![ezgif com-gif-maker (8)](https://user-images.githubusercontent.com/52202474/121065573-dfcad080-c803-11eb-9861-f6825d5ea819.gif)

### 회원가입
![ezgif com-gif-maker (9)](https://user-images.githubusercontent.com/52202474/121065579-e22d2a80-c803-11eb-9867-1297a05f8d54.gif)

### 역, 노선, 구간 추가 삭제기능
![ezgif com-gif-maker (10)](https://user-images.githubusercontent.com/52202474/121065578-e1949400-c803-11eb-8ef5-1ff422ac88ad.gif)

## 구현 기능 목록

### 회원 기능

- [x] 사용자는 회원 가입을 할 수 있다.
- [x] 사용자는 가입한 계정으로 로그인을 할 수 있다.
- [x] 사용자는 로그인하여 지하철 노선도 관리 페이지에 접근할 수 있다.

### 지하철 역 관리 기능

- [x] 사용자는 등록되어 있는 전체 지하철 역 목록을 조회할 수 있다.
- [x] 사용자는 지하철 역을 추가할 수 있다.
- [x] 사용자는 지하철 역을 삭제할 수 있다.

### 지하철 노선 관리 기능

- [x] 사용자는 등록되어 있는 전체 지하철 노선 목록을 조회할 수 있다.
- [x] 사용자는 지하철 노선을 추가할 수 있다.
- [x] 사용자는 등록되어 있는 지하철 노선을 삭제할 수 있다.

### 지하철 구간 관리 기능(노선에 상행역,하행역 외에 중간역 추가 기능)

- [x] 사용자는 특정 노선의 전체 구간 목록을 확인할 수 있다.
- [x] 사용자는 노선에 등록되어 있는 구간을 삭제할 수 있다.

## 궁금한 점

### 에러, 로딩처리
각 요청에 대한 에러와, 로딩 상태를 처리하는 부분이 쉽지가 않네요!
각 페이지마다 반복적으로 구현하고 싶지 않아 App에서 에러메시지와, 로딩화면을 보여주려고 시도하였습니다.
에러상태를 에러발생 후 2초 뒤에 초기화 되도록 설정해보았습니다. 
좀더 효율적인 에러, 로딩 처리방법이 궁금합니다. 

### 테스트
로그인, 역 관리 페이지에 대해서만 테스트를 진행하였습니다. 다른 테스트도 구현했어야하는데 시간이 너무 모자르네요..
각 페이지에서 input에 값을 입력하고, submit 버튼을 클릭하였을 때, 삭제버튼을 눌렀을 때  올바른, 올바르지 않은 액션이 dispatch 되는지 까지를 검사하였습니다.
시간이 좀 더 있었다면, 각 페이지에 대하여 통합테스트를 진행한 후 reducer, customhook에 대한 단위테스트를 진행하면 좋겠다고 생각이 들었습니다.
정해진 기간안에 효율적으로 프로젝트를 진행해야 한다고 생각했을 때, 광휘님이 선호하시는 또는 우선순위로 생각하시는 테스트 
과정을 공유해주시면 좋겠습니다!

### 커스텀 훅
폼과 인풋이 많다보니 useInput(validator)와 같은식으로 호출하고, 관련 메서드와 값을 반환합니다.
onChange 에서 값을 변경하면서 유효성을 검사하고 그에 대한 에러메시지 상태 값을 hook에서 반환하는데 적절히 이루어졌는지 더 개선할 방법이 있을지도 궁금합니다.
역, 라인 / 구간단위로 커스텀훅을 적용하여 CRUD로직을 분리하였는데 적절한 규모인지 궁금합니다. 좀더 개선할 방법이 있는지도 궁금합니다.🤔
 
**광휘님의 조언이 시급합니다! 😅**

swon3210 committed:
chore: 프로젝트 초기 환경 세팅

swon3210 committed:
docs: 기능 목록 작성

seojihwan committed:
chore: styling 환경 세팅

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: Global 스타일 적용

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: Modal 컴포넌트 구현 및 스토리북 작성

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: Confirm 컴포넌트 구현 및 스토리북 작성

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
chore: Confirm, Modal 컴포넌트 story title 변경

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: Button 컴포넌트 구현 및 스토리 작성

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: select 컴포넌트 구현 및 스토리 작성

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
refactor: 스타일 관련 리팩토링

- 인터페이스 네이밍 변경
- 스타일 컴포넌트 네이밍 변경
- li 태그의 list-style 제거
- 버튼 스토리 변경

Co-authored-by: SONG WON LEE <swon321@gmail.com>

seojihwan committed:
feat: ListItem 컴포넌트 구현 및 스토리 작성

- 공통 스타일 컴포넌트 구현

Co-authored-by: SONG WON LEE <swon321@gmail.com>

swon3210 committed:
feat: Input 컴포넌트 구현 및 스토리 작성

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: Block 컴포넌트 구현 및 스토리 작성

swon3210 committed:
feat: ScrollArea 컴포넌트 구현 및 스토리 작성

swon3210 committed:
feat: LabeledNode 컴포넌트 구현 및 스토리 작성

swon3210 committed:
feat: 페이지 라우팅 적용

swon3210 committed:
refactor: 페이지 컴포넌트 폴더명 변경

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: Header 컴포넌트 구현 및 스토리 작성

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
refactor: ThemeProvider, GlobalStyle 위치 변경

swon3210 committed:
refactor: Button 컴포넌트 글자 상하 가운데 정렬 적용

swon3210 committed:
feat: 페이지 컴포넌트에 flex center 적용

swon3210 committed:
feat: 컴포넌트 스타일링 오류 수정

swon3210 committed:
feat: 전역 스타일링 수정

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: size 상수 모듈화, route 상수 생성

swon3210 committed:
fix: 네비게이션 버튼 레이아웃 마진 수정

swon3210 committed:
fix: ListItem 컴포넌트이 기본 attribute prop을 할당 받도록 수정

swon3210 committed:
style: App 컴포넌트 return 문 제거

swon3210 committed:
feat: heading 관련 css normalize 추가

swon3210 committed:
feat: ListItem 컴포넌트 스타일링 변경

swon3210 committed:
feat: StationManagementPage 컴포넌트 구현

swon3210 committed:
fix: route 에 exact 속성 부여

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: 지하철 노선색 부여를 위한 색상 객체 생성 및 적용

swon3210 committed:
feat: LineManangementPage 컴포넌트 작성

swon3210 committed:
fix: Select 컴포넌트가 options 을 prop 으로 받을 수 있도록 변경

swon3210 committed:
feat: SectionManagementPage 컴포넌트 구현

swon3210 committed:
fix: 구간관리 이모지 출력 오류 수정

swon3210 committed:
feat: LoginPage & signupPage 컴포넌트 구현

swon3210 committed:
chore: favicon 변경

swon3210 committed:
chore: index.html 의 필요없는 코드 삭제 & 제목과 설명 명시

swon3210 committed:
fix: Button 컴포넌트에 type 명시 & Login, Signup 페이지에 form 태그 추가

swon3210 committed:
feat: ColorPicker 컴포넌트 구현 및 스토리 작성

swon3210 committed:
refactor: LineManagementPage 컴포넌트 레이아웃 수정

swon3210 committed:
feat: LineAddModal 컴포넌트 구현

swon3210 committed:
feat: LineAddModal 컴포넌트에 시간 Input 제거 & 입력 제한 attribute 추가

swon3210 committed:
feat: Select 컴포넌트에 default option 설정 기능 추가

swon3210 committed:
feat: SectionAddModal 컴포넌트 구현

swon3210 committed:
feat: 리덕스 초기 설정

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: auth 리듀서 구현

swon3210 committed:
feat: Input 관련 커스텀 훅 구현

- Input 이 에러메세지로 null 도 받을 수 있도록 수정
- age, email, password 관련 유효성 검증 함수 작성

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: 로그인 & 회원가입 미들웨어 구현

- 리덕스 초기화를 위해 station 리듀서 초기화

swon3210 committed:
feat: 로그인 & 회원가입 기능 구현

swon3210 committed:
feat: 로그인 & 회원가입시 리다이렉션 기능 구현

- 로그인 & 회원가입 input required 설정
- 로그인 & 회원가입 input type 설정
- 회원가입 나이 input min, max 설정

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
feat: 새로고침 시 로그인 상태 유지 기능 구현

- 로그인 되어 있지 않는 경우 로그인 페이지 외 다른 페이지로 넘어갈 수 없도록 처리

swon3210 committed:
docs: 허용가능 색상 수 10 -> 100으로 증가

swon3210 committed:
feat: 로그아웃 기능 구현

swon3210 committed:
feat: 역 조회 & 추가 기능 구현

swon3210 committed:
docs: 기능 구현 현황 체크 및 요구 사항 수정

swon3210 committed:
feat: 역 삭제 기능 구현

swon3210 committed:
refactor: useInput 훅이 배열이 아닌 객체를 리턴하도록 수정

swon3210 committed:
refactor: station 훅 파일 이름을 useStation 으로 변경

swon3210 committed:
refactor: Auth 관련 타입 위치 변경

swon3210 committed:
refactor: 기존 dispatch 호출 의존성을 커스텀 훅에서만 부여되도록 수정

swon3210 committed:
feat: 노선 추가 & 삭제 기능 구현

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
fix: Select 컴포넌트의 기본 옵션이 value 를 가지지 않은 채 selected 되어 있도록 수정

Co-authored-by: 서지환 <seojihwan213@naver.com>

swon3210 committed:
refactor: useSelect 훅이 selectValue 상태의 기본값을 반드시 넘겨받도록 수정

swon3210 committed:
feat: 구간 추가 및 삭제 기능 구현

swon3210 committed:
fix: Select 컴포넌트가 default value = "" 를 기본으로 선택하도록 수정

swon3210 committed:
feat: API 변경 기능 구현

swon3210 committed:
feat: 데이터 추가 및 삭제 이후 다시 데이터 조회를 통해 요청사항이 반영되도록 로직 수정

swon3210 committed:
feat: 잘못된 입력값 & 요청에 대한 유효성 검사 구현

- alert 메세지로 사용자에게 피드백

swon3210 committed:
test: 테스트용 데이터 mock 파일 생성

swon3210 committed:
docs: 비슷한 기능 목록 합치기

swon3210 committed:
test: StationManagementPage 테스트 코드 작성

swon3210 committed:
test: StationManagementPage 테스트 작성

Co-authored-by: 서지환 <seojihwan213@naver.com>

seojihwan committed:
refactor: input값 유효성검사를 change 이벤트에서 하도록 수정

seojihwan committed:
fix: Select컴포넌트 스토리의 파라미터 수정

seojihwan committed:
test: 중복된 셋팅 코드 setupTest로 이동

- axios, useSelector
- cleanup
- window.alert

seojihwan committed:
refactor: 주석제거, 불필요한 import 제거

seojihwan committed:
refactor: Navigation를 컴포넌트로 분리

seojihwan committed:
refactor: route들 App에서 분리

seojihwan committed:
refactor: 요청 실패시 alert 대신 메시지를 화면에 출력

seojihwan committed:
refactor: 페이지 너비 상수화, import 순서 변경

seojihwan committed:
refactor: 컬러 enum으로 변경, 모듈 import 순서 변경

seojihwan committed:
docs: README.md 수정

- 기존 README.md와 요구사항과 REQUIREMENTS.md를 종합한다

seojihwan committed:
test: query의 우선순위에 맞게  Role과 Text query사용하기

- data-testid삭제

seojihwan committed:
test: 로그인 페이지 테스트

- 이메일, 비밀번호 길이와 형식에 대하여 실패 케이스와 성공케이스를 작성한다.
- accessToken값을 반환하도록 axios.post 모킹값을 수정한다.

seojihwan committed:
test: 모듈 import 순서, 값을 명시적으로 보여줄 수 있도록 수정

seojihwan committed:
feat: 로딩 컴포넌트 추가

seojihwan committed:
fix: 로딩 상태 추가

seojihwan committed:
test: only 제거, 초기 상태값에 loading 추가

seojihwan committed:
refactor: 에러 메시지, 로딩 컴포넌트 위치 수정

- 에러 메시지 컴포넌트 추가

seojihwan committed:
fix: 사용되지 않는 타입 제거

seojihwan committed:
reafctor: 로딩 컴포너트 기본 색상, 위치 변경

seojihwan committed:
refactor: 사용되지 않는 style파일 삭제, 높이 상수화

- 각 Page의 스크롤 추가, 높이에 대한 상수 값 적용

seojihwan committed:
refactor: 사용되지 않는 코드 삭제

seojihwan committed:
reafctor: 역, 노선, 구간 삭제버튼 클릭시 컨펌모달로 재 확인

seojihwan committed:
test: confirm 버튼에 data속성 부여, <div id="modal"></div>이 없는경우 생성

seojihwan committed:
refactor: matcher를 이용하여 action에 대한 중복 처리 제거

- thunkAction의 ( pending, fulfilled, rejected )상태에 대한 type guard 유틸 함수들 생성

seojihwan committed:
refactor: 컴포넌트 re-export적용

- 중복 코드 제거
- components/@shared => components/Layout

seojihwan committed:
fix: slice에 전달되는 액션에 대한 필터링 추가

- 액션 prefix ( [AUTH], [STATION], [LINE] ) 에 대한 필터링

seojihwan committed:
refactor: hook re-export 적용

- 순환 참조 제거

seojihwan committed:
feat: 에러 초기화 기능 추가

seojihwan committed:
refactor: 상수, 페이지 re-export 적용

seojihwan committed:
refactor: 파일 이름 변경 types.ts -> index.ts

seojihwan committed:
refactor: import export 위치 통일하기

- export는 하단에 모아준다.
- type은 export type을 사용한다

seojihwan committed:
docs: 기능 요구사항 확인

seojihwan committed:
fix: 스토리북 ThemeProvider 제거

seojihwan committed:
refactor: 코드 스타일에 맞게 수정

seojihwan committed:
docs: 적용되지 않은 기능목록 체크해제

Vallista commented on package.json:
```diff
@@ -0,0 +1,83 @@
+      "not op_mini all"
+    ],
+    "development": [
+      "last 1 chrome version",
```
브라우저리스트가 무엇일까요?
요걸 의도하고 작성하신게 맞으실까요?

seojihwan replied:
cra 템플릿을 사용했기 때문에 의도하고 작성한 설정은 아닙니다! 😅

broswerslist정보를 통해 개발 환경과 배포 환경에서의 자바스크립트 지원 문법, 브라우저 기능을 파악하고 Babel transform plugins과  core-js를 통해 코드를 변환해주는 것으로 알고 있습니다.
>We leverage these data sources to maintain mappings of which version of our supported target environments gained support of a JavaScript syntax or browser feature, as well as a mapping of those syntaxes and features to Babel transform plugins and core-js polyfills.

지원 범위를 살펴보니 es6을 지원하지 않는 Opera Mini등의 브라우저를 browserslist에서 제외하여  빌드시간과 번들파일의 크기를 줄일수 있도록 설정되었네요!
![image](https://user-images.githubusercontent.com/52202474/122389337-ccf68f80-cfab-11eb-85aa-722c90124f1a.png)

![image](https://user-images.githubusercontent.com/52202474/122389364-d54eca80-cfab-11eb-8ec4-93cc74ee6486.png)
2년전의 자료이지만 지원 브라우저에 따라 대략 15%정도의 번들파일의 사이즈 차이가 발생할 수 있는것으로 보입니다!

Vallista commented on src/@types/index.ts:
```diff
@@ -0,0 +1,62 @@
+}
+
+interface Section {
+  upStation: {
```
위의 interface Station 을 쓰면 되지 않나요?

seojihwan replied:
바로 위의 타입을 재활용하면 되는데 이걸 놓쳤네요😅 수정했습니다! b0121b6b0e37aa631e553259cf916dee7409a5f9

Vallista commented on src/App.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  return (
+    <BrowserRouter>
+      <Header style={{ marginTop: "1.5625rem", marginBottom: "1.5625rem" }}>
```
인라인 스타일을 사용할 것이면, emotion을 써서 `css=?` 형태로 사용해보는 것은 어떨까요?

Vallista replied:
아니 그 이전에 styled-component에는 없던가..?

seojihwan replied:
인라인스타일을 사용하여 기존 컴포넌트에서 확장된 스타일을 적용하고 스타일 적용에 있어 우선순위를 보장하고 싶었습니다만 좋지 않은 선택이었네요! 😭
class보다 성능도 좋지 않고, 정적인 style을 굳이 inline으로 부여할 필요가 없다는 것을 알게되었습니다!

> Some examples in the documentation use style for convenience, but using the style attribute as the primary means of styling elements is generally not recommended. In most cases, className should be used to reference classes defined in an external CSS stylesheet. style is most often used in React applications to add dynamically-computed styles at render time. See also FAQ: Styling and CSS.

babel-plugin-styled-components을 이용하면 css prop을 class로 만들어 마지막에 추가해주기 때문에 같은 레벨에서의 스타일 우선순위도 보장받을 수 있네요  !!
> NOTE
To enable support for the css prop you have to use the Babel plugin.

92fe99cc27c1506c2953ccf74c02c3d335550683
4dc82777f06f391245a716d98dbcc7ce643e3ed9

Vallista commented on src/Routes.tsx:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <>
+      <PublicRoute exact path={PAGE_PATH.LOGIN}>
```
```tsx
interface RouteShape {
  private: boolean
  exact: boolean
  path: string | string[]
  component: React.ReactNode
}

const routes: RouteShape[] = [
  {
    private: false,
    exact: true,
    path: '/signup',
    component: Signup
  },
  {
    private: true.
    exact: true,
    path: ['/home', '/station_management'],
    component: Signup
  }
  ...
]
```

```tsx
return (
  <>
    {
       routes.map(({private, component: ChildComponent, ...otherProps }) => {
         const TargetElement = private ? PrivateRoute : PublicRoute
         
         return (
           <TargetElement {...otherProps}>
             <ChildComponent />
           </TargetElement>
         )
       })
    }
  </>
)
```

참고, 컴파일없이 생각나는대로 짠 로직이라 안될 수 있음.

쨋든 요런 식으로 추상화하는게 더 낫지 않을까요?

Vallista replied:
생산성이란, 수정하기 쉬워야 하며 한 곳에서 관리가 가능해야 합니다. (관리의 포인트가 줄수록 좋습니다.)

현재 코드는 PAGE_PATH를 관리해야하는 문제를 갖고있고, PAGE_PATH에 추가하고, Routes.tsx에 와서 다시 수정해야하는 문제가 있죠. 위처럼 코딩하면 더 깔끔쓰~

seojihwan replied:
동의합니다!! 관리 포인트도 한곳으로 줄고, 중복코드를 줄일 수 있어 훨씬 보기 좋아지네요~👍 73096cc02d0c4216d88b82ce47fb3c34534546e0

Vallista commented on src/apis/index.ts:
```diff
@@ -0,0 +1,22 @@
+import { requestLine } from "./line";
+import { requestSection } from "./section";
+
+enum BASE_URL {
```
이 코드는 전 개인적으로 오버엔지니어링이라고 생각해요.
enum을 쓰는 이유를 이 상황에서 보았을 때 3가지가 있어 보입니당.

1. 열거를 하기 위함 (벨류가 특수히 길거나 아에 연관이 없어, 키로 맥락을 나타내야 할 경우)
2. 컴파일 시점에서 열겨된 값의 타입 체킹 혹은 그러한 것들이 필요한 경우
3. 역 매핑이 필요한 경우

첫 번째는, 객체지향 관점에서 저는 동의합니다.
enum은 이러한 키-값에 대한 요구사항에 대다수의 프로그래밍 언어에서 효과적인 해결책이거든요.
두 번째는 이 상황에서는 필요가 없군요.
세 번째도 이 상황에서 필요가 없습니다. (일단 역매핑을 쓰고는 있으나,,, 아래서 서술)

그렇기 때문에 비단 열거형태만 가져오기 위해 사용하게 됩니다. 그렇다면 사실, 아래가 더 유효합니다.

```ts
const BASE_URL = {
  에드: "https://subway-app.kro.kr",
  수리: "https://suri-subway.kro.kr",
  와일더: "https://wilder-subway.kro.kr",
  포모: "https://subway-pomo.kro.kr",
}
```

왜 이게 더 유효할까요? 그렇다면 조금 더 자세히 들여다봅시다.

ts가 컴파일되면 현재 코드는 아래와 같이 됩니다.

```tsx
"use strict";
var BASE_URL;
(function (BASE_URL) {
    BASE_URL["\uC5D0\uB4DC"] = "https://subway-app.kro.kr";
    BASE_URL["\uC218\uB9AC"] = "https://suri-subway.kro.kr";
    BASE_URL["\uC640\uC77C\uB354"] = "https://wilder-subway.kro.kr";
    BASE_URL["\uD3EC\uBAA8"] = "https://subway-pomo.kro.kr";
})(BASE_URL || (BASE_URL = {}));
const changeBaseUrl = (범인) => {
    // localStorage.setItem("범인", 범인);
};
```

무언가 길죠? 그렇다면 단순한 const 형태로 봐봅시다.

```ts
"use strict";

const BASE_URL = {
    에드: "https://subway-app.kro.kr",
    수리: "https://suri-subway.kro.kr",
    와일더: "https://wilder-subway.kro.kr",
    포모: "https://subway-pomo.kro.kr",
};
const changeBaseUrl = (범인) => {
    // localStorage.setItem("범인", 범인);
};
```

더 간단하죠? 심지어 원래 형태로 뽑아주네용???

자 여기서 어떤 것을 알 수 있냐.. enum은 아까 위에서 언급한 3가지 중, 3번째를 지원합니다. 바로 역매핑입니다. 역매핑은 무엇일까요? 바로 아래와 같은 것을 이야기합니드아..

```ts
const BASE_URL = {
  에드: "https://subway-app.kro.kr",
  수리: "https://suri-subway.kro.kr",
  와일더: "https://wilder-subway.kro.kr",
  포모: "https://subway-pomo.kro.kr",
}

let a = BASE_URL.포모
BASE_URL[a] // '주소'
```

즉, 아래의 현재 코드에서 BASE_URL은 역매핑만을 쓰고 있습니다. 정상적인 맵핑을 쓰고 있지 않쥬.. 그럼 과연 이 친구는 필요한 친구일까요? 일반적으로 사용되는 프로그래밍 관점에서의 enum을 쓰고 있는 걸까요? 비단, 역 매핑을 위해서 혹은 가독성을 위해서 이 코드는 두 가지를 희생했습니다.

1. const로 나타내는 것보다 훨씬 더 긴 빌드시의 코드
2. [웹팩 및 번들러에서의 트리셰이킹 문제](https://engineering.linecorp.com/ko/blog/typescript-enum-tree-shaking/) (물론 저는 이 글에서 제시하는 트리셰이킹 문제에 대해서 감수할 수 있다 생각하며, 크게 문제가 있다고 생각하지 않습니다. 오히려, enum을 쓰는 쪽을 좋아합니다.)

어찌 되었던, 제일 중요한 것은 const만으로 해결할 수 있는데 enum을 강제로 사용한 느낌... 이라는 것입니다.. 네..

seojihwan replied:
enum에 대한 이해도가 부족했네요 ㅎㅎ
단순히 값을 재 할당하는 것을 막기위해 사용했는데 말씀해주신 enum의 장점을 활용하지 못하는 상황이네요!

enum 대신 object as const를 쓰면 빌드시의 더 긴 코드로 변환되는 단점을 피하고, 재 할당을 막는 장점을 갖을 수 있겠네요!
db6318724daeca912beca24abe36d0a65cc2f41a

>  Objects vs Enums
In modern TypeScript, you may not need an enum when an object with `as const` could suffice

tree shaking에 점은 아래 코멘트의 COLOR값을 enum에서 object as const로 변경하면서 더 깨달을 수 있었습니다!

Vallista commented on src/apis/line.ts:
```diff
@@ -0,0 +1,23 @@
+
+import { Line, LineAddRequestItem } from "../@types";
+
+const requestLine = {
```
호호.. 이 apis에는 수많은 문제가 있습니다...

1. 에에엥? 이미 `apis/line.ts`는 자신을 충분히 잘 나타내고 있는 이름입니다! 그렇기 대문에 requestLine이라는 래핑이 필요없어 보입니다!! (그런데 restful 하게 이름도 lines.ts 면 좋겠네용.)
2. get이 실패할 수도 있는데 무엇을 믿고 성공값만 bypass 하나요..?
3. try-catch 어디갔으아아아아악!@ (화남)

```tsx
type HTTPResponseCode = '401'

export enum ErrorMessage {
  ERROR401 = '에러 메시지 테스트',
}

interface ApiSuccessShape<T> {
  success: true
  result: T
}

interface ApiFailureShape<T extends string> {
  success: false
  code: `${number}`
  result: T // 에러 메시지가 보통 내려올 것이므로 extends string 형으로 받음.
}

type ApiResponse<T> = ApiSuccessShape<T> | ApiFailureShape<ErrorMessage>

export default {
  async get(): Promise<ApiResponse<Line[]>> {
    try {
      const response = await axios.get<Line[]>("/lines");
      return {
        success: true,
        result: response.data
      }
    } catch (err) {
      const code = err.response.code as HTTPResponseCode
      return {
        success: false,
        code,
        result: ErrorMessage[`ERROR${code}` as keyof typeof ErrorMessage]
      }
    }
  }
}
```

좀 더 안전하게... 그리고 타입추론이 가능하도록 짜는게 좋겠습니다.

Vallista replied:
위 예제에서는 type alias로 interface union을 만들어 success의 true, false로 타입에 따라 다른 추론이 되도록 구현한 예제입니다. 도움이 될 터이니 기억해두세영~

seojihwan replied:
변명을 하자면 dispatch(action)을 할때 createThunkAction에서 에러 처리를 하고 있었지만 타입 추론은 이루어지지 않고 있었네요ㅠㅠ

제시해주신 방법대로 api요청함수에서 에러처리를 하고 그 함수를 외부에서 사용하는 것이 더 일관성있는 코드를 가져가는데 좋을것 같습니다.

제시해주신 코드를 바탕으로 api요청하는 함수에 대하여 추상화하고 이를 사용하도록 구성해보았습니다!!
```ts
type Method = "get" | "post" | "put" | "delete";

interface options {
  method?: Method;
  data?: any;
  config?: AxiosRequestConfig;
}

interface SuccessShape<T> {
  success: true;
  result: T;
}

interface FailureShape<T extends string> {
  success: false;
  code: string;
  result: T;
}

type Response<T> = SuccessShape<T> | FailureShape<string>;

const request = async <T>(
  path: string,
  { method = "get", data, config }: options = { method: "get" }
): Promise<Response<T>> => {
  try {
    const response =
      method === "get" || method === "delete"
        ? await axios[method]<T>(path, config)
        : await axios[method]<T>(path, data, config);

    return {
      success: true,
      result: response.data,
    };
  } catch (error) {
    return {
      success: false,
      code: error.response.code,
      result: error.response.data,
    };
  }
};
```
그리고 반환되는 값을 이용하여 createAsyncThunk에서 액션에 대한 결과를 처리할 수 있었습니다!

api요청에서 타입추론이 쉽지 않았는데 이번기회에 많은 경험치를 얻을 수 있었네요~ 감사합니다 ㅎㅎ👍👍👍

d9e45ff7c8b0d7b0d098c033c64f300f36e4a758

Vallista commented on src/apis/section.ts:
```diff
@@ -0,0 +1,35 @@
+
+import { SectionAddRequestItem } from "../@types";
+
+const requestSection = {
```
어차피 apis/section 안에 있으므로 add 라는 것은 section 추가하는 api를 호출한다와 동일합니다.
위 피드백처럼 동일하게 해주세영~

seojihwan replied:
불필요하게 긴 네이밍을 사용했네요 ~ 수정했습니다!

Vallista commented on src/components/Button/Button.styles.ts:
```diff
@@ -0,0 +1,65 @@
+  buttonTheme?: "primary" | "gray" | "white";
+}
+
+type ButtonStylesProps = ButtonBlockProps;
```
?_? 이걸 왜 한번 더 래핑하는걸까요?

Vallista replied:
양식이 interface union이라 이런걸까요??

seojihwan replied:
컴포넌트를 구성하는 styled component들의 props에 대한 interface를 묶어주기 위해 사용했습니다.

type StyleProps = PropA | PropB 요런식으로 사용해볼려고 했습니다.
```ts
export type Props = ButtonHTMLAttributes<HTMLButtonElement> & AstyleProp & BstyleProp;

const Button = (props: Props) => <ButtonBlock {...props} />;
```
버튼의 속성에 대한 타입과 각각의 스타일관련 Prop의 타입에 대한 중요도가 같지 않다고 생각하여 위의 상황을 피해보고 싶었습니다. ( 실제로 일어나지 않음 )

컴포넌트에서는 이렇게 사용하고 있는데요.
```ts
export type Props = ButtonHTMLAttributes<HTMLButtonElement> & ButtonStylesProps;

const Button = (props: Props) => <ButtonBlock {...props} />;
```

처음에는 코드 스타일을 일관되게 가져가는데 도움이 되지 않을까 생각했었는데 실제로 스타일 관련Prop이 여러개 사용되는 경우도 없었고 전자의 상황이 더 코드를 파악하는데 좋을것 같네요! 수정하였습니다.😅 4be08796e5fdd7e071703634ecefe1b4065d701a

Vallista commented on src/components/Button/Button.styles.ts:
```diff
@@ -0,0 +1,65 @@
+
+interface ButtonBlockProps {
+  kind?: "rect" | "eclipse";
+  size?: "sm" | "md" | "block";
```
요렇게 사이즈를 주고 컴포넌트를 구성하면 이 것은 추후 생산성을 떨어뜨리는 양날의 검이 될텐데.. 흠 경험해보시는 것을 추천..

seojihwan replied:
컴포넌트의 크기를 size = "30"등으로 좀 더 유연하게 구성할 수 있도록 만들면

사용하는 입장에서 일관성있는 사이즈 등의 옵션을 선택하기 어려울 것 같다고 생각됩니다. 

물론 상수로 하나의 프로젝트에서 버튼의 크기 등등을 관리할 수 있겠지만? 다른 프로젝트에도 같은 컴포넌트를 사용할때 서비스 이용자에게 일관된 경험을 제공하기 어려울 것 같다고 생각하였습니다.
따라서 유연함을 더 살리기 보다는 제한된 방식으로 컴포넌트를 사용할 수 있도록 하였습니다.

Chakra UI나 Ant Design에서도 Input 컴포넌트를 살펴본 결과 size를 large | middle | small 이런식으로 사용하는 것으로 보입니다.

이에 대해도 어떻게 생각하시는지 궁금합니다!! 예상 되는 어려움을 좀더 공유해주시면 큰 도움이 될 것 같습니다 😭

Vallista commented on src/components/Confirm/Confirm.styles.ts:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  > button:hover {
+    background-color: #dadada;
```
#dadada 같은 애들을 constants로 빼놓는게 낫지 않을까요?

seojihwan replied:
이전 미션에 사용한 컴포넌트를 고대로 가져와서 사용했는데 상수화를 안했네요 😅
모든 컬러 값들을 찾아서 상수값을 적용하였습니다! 931cfcd0c81f4922b202c01328b7bf56c20c0b8e

Vallista commented on src/components/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,33 @@
+`;
+
+const Contents = styled.div`
+  z-index: 1004;
```
z-index도 중앙관리가 되었으면 합니다~

seojihwan replied:
z-index도 상수화 해서 관리포인트를 모아주는 것이 좋겠네요~~ 감사합니다👍👍 4f7f7f0ff59f49eb23ae4038baebf91272bec794

Vallista commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,44 @@
+import { ModalBlock, Dimmed, Contents } from "./Modal.styles";
+
+export interface Props {
+  onClose: MouseEventHandler<HTMLDivElement>;
```
Modal내에서 상태를 캡슐화하여 갖고있고, onClose가 외부에 dependency되면 안됩니다!

React.createPortal로 돔 최하단에 그려주되, 닫기 버튼을 클릭하면 내부에서 상태를 변경하도록 하고, 삭제하면서 사이드 이펙트로 다른 이벤트가 실행될 수 있도록 일종의 파이프라인만 props로 뚫어주는게 좋겠습니다!

Vallista commented on src/components/index.ts:
```diff
@@ -0,0 +1,49 @@
+  Grid,
+} from "./Layout";
+
+export {
```
이렇게 다 export 하면 번들링 과정에서는 어떤 결과가 일어날까용?

seojihwan replied:
질문에 대한 해답을 찾지 못했습니다.😭

직접 빌드를 해보았는데 빌드 시간의 차이도 유사하고, 빌드된 코드는 확인해본 결과 일치했습니다.

빌드 이전의 코드입니다. 

하위 폴더에서 모아서 re-export / re-export하지 않고 entry파일인 index.js에서 직접 import
![image](https://user-images.githubusercontent.com/52202474/122421781-99296300-cfc7-11eb-8e62-1cf36b57bb3c.png)

빌드 이후의 코드입니다.
![image](https://user-images.githubusercontent.com/52202474/122421949-b6f6c800-cfc7-11eb-8e66-e6e395f1e3f0.png)
![image](https://user-images.githubusercontent.com/52202474/122422051-cb3ac500-cfc7-11eb-991e-44a08126f835.png)

어던 차이가 있는지 잘 모르겠습니다. 🤔
힌트나 알아보면 좋을 키워드를 알려주실 수 있나요??

Vallista commented on src/constants/color.ts:
```diff
@@ -0,0 +1,216 @@
+enum COLOR {
```
와우;;;

Vallista commented on src/constants/color.ts:
```diff
@@ -0,0 +1,216 @@
+  PINK_900 = "rgb(131, 24, 67)",
+}
+
+enum CIRCLE_COLOR {
```
ㅋㅋㅋㅋㅋㅋㅋㅋ 너무 디테일하게 한거 아닌가요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

Vallista replied:
적당히 쓰는값만 하면 좋을텐데!

Vallista replied:
그리고 써클컬러는 없어도 되지 않을까나..

seojihwan replied:
API 명세가 작성된 예시에서는 케밥 케이스 형식으로 color에 대한 값이 예시로 주어졌습니다!

예시 데이터
![image](https://user-images.githubusercontent.com/52202474/122412097-413b2e00-cfc0-11eb-8a81-c61cc96a00c9.png)

그래서 이에 대한 값(bg-red-600)을 key 로 관리하고, 대응 되는 rgb 컬러를 제공하기 위해 써클 컬러를 객체로 관리하였습니다. 

다른 백엔드 크루들과 제 페어도 이와 같은 값으로 테스트하고 있었기 때문에 파스칼 케이스를 쓰고 싶었지만 케밥 케이스를 사용하였습니다!

써클 컬러에는 키 값에 bg라는 네이밍(bg-red-600)이 들어감으로써 background color에 한정되어 사용가능하다는 생각이들어.. 중복이지만 color라는 객체를 만들어서 사용하였습니다!!

기존에 사용했던 enum의 경우에는 사용하지 않고 있던 색상에 대한 코드도 모두 번들 파일에 포함되었습니다.
![image](https://user-images.githubusercontent.com/52202474/122412430-852e3300-cfc0-11eb-8932-fa93ebcf586e.png)

enum들의 장점을 살릴 수 없다고 판단하여 object as const로 대체하여 사용하였고 tree shaking문제도 해결할 수 있었습니다.
![image](https://user-images.githubusercontent.com/52202474/122412409-7d6e8e80-cfc0-11eb-8a3e-d2262e680c00.png)
위의 번들파일 처럼 사용되는 컬러값들만 번들 파일에 포함되고 있기 때문에 사용하지 않는 컬러들은 삭제하지 않았습니다.
다음 단계 미션을 진행하면서 사용될 수도 (?) 있을 것같아 두는 것이 편의성이 높다고 생각됩니다!

써클 컬러들은 동적으로 아래의 컴포넌트에서 사용될 수 있기 때문에 tree shaking이 되지 않더라도 남겨두어야 할 것 같습니다 ㅠㅠ
![image](https://user-images.githubusercontent.com/52202474/122413400-4cdb2480-cfc1-11eb-826b-eb05ad1ac0b4.png)

Vallista commented on src/hooks/useLine.ts:
```diff
@@ -0,0 +1,59 @@
+
+import { LineAddRequestItem, SectionAddRequestItem } from "../@types";
+
+const useLine = () => {
```
아무리 생각해봐도 useLine, 등의 dispatcher hooks는 잘 맹글었네요 ㅋㅋ 억까 하려고 계속 봤는데 없넹..

Vallista commented on src/hooks/useLine.ts:
```diff
@@ -0,0 +1,59 @@
+    await dispatch(action.getLines());
+  };
+
+  const addLine = async (lineRequestItem: LineAddRequestItem) => {
```
다만 에러로직이 addLine api 내에 없다면, 여기에라도 있었어야 해용

Vallista commented on src/modules/station.ts:
```diff
@@ -0,0 +1,125 @@
+import {
```
으음.. 쭉 살펴봣지만 RTK는 적절히 사용한 것으로 보입니다만.. 전체적으로 코드가 산발적이네요.
ducks 패턴을 쓰므로, 해당 모듈 안에

```
modules
 ㄴ station
       ㄴ index.ts (reducer)
       ㄴ thunks.ts (thunks)
       ㄴ types.ts (typescript interface, and so on..)
```

모듈화점 해주세영~

seojihwan replied:
제가봐도 읽기 좋은 코드가 아니네요 ㅠㅠ😭 폴더구조를 변경하였습니다!!!
631b2bf7f24b1b05f2339c639a085abf41842ca8

Vallista commented on src/pages/LineManagement/Modal/LineAddModal.tsx:
```diff
@@ -0,0 +1,143 @@
+import { CIRCLE_COLOR } from "../../../constants";
+import { LineAddRequestItem, Station } from "../../../@types";
+
+interface Props {
```
모달들이 너무 제각각 공통된 로직을 갖고있어용~

Vallista replied:
이 로직은 위에서 말했듯, 공통 모듈로 축약하고 상태 캡슐화하면서 어느정도 보완될 것으로 보입니다용~

Vallista commented on src/pages/Login/LoginPage.test.tsx:
```diff
@@ -0,0 +1,186 @@
+
+describe("로그인 페이지 테스트", () => {
+  describe("사용자가 유효한 비밀번호(4글자 이상 20글자 이하의 길이)를 입력한 상황에서", () => {
+    describe("사용자가 입력한 이메일이 30글자이하의 유효한 형식인 경우", () => {
```
e2e 에서 원하는 시나리오 테스트 굿

Vallista commented on src/pages/Login/LoginPage.test.tsx:
```diff
@@ -0,0 +1,186 @@
+          </Provider>
+        );
+
+        fireEvent.change(getByPlaceholderText(/이메일/i), {
```
음 이 로직들 뭔가 반복되는.... 줄일 수 있겠는데요?

seojihwan replied:
beforeEach를 이용하여 컴포넌트 렌더링 관련 코드의 중복을 제거하였습니다~👍

3b6d7862fcf1c1e4be124f3beb07d18003f2f417

Vallista reviewed (changes requested):
안녕하세요 브랜~
오랜만에 뵙지만 이제 곧 긴 안녕을 하겠군여
최선을 다해 리뷰했습니당~

# UX 피드백

코드에 집중하시라고, 진행하지 않았습니다!

# 전반적인 코드 피드백

1. 모달쪽 중복 코드가 너무 많습니다~ 이 문제는 모달이 제대로 캡슐화가 안되었고, 상태를 외부에 위탁하기 때문인데요~
해당 부분은 추상화를 통해 축약이 가능해보입니다! ModalManager 같은것을 만들어봐도 좋겠어요~
Context로 모달이 쓰일수 있는 전역 Provider를 만든 후, 내부에서 어디든지 hooks를 호출하여 createModal 하면나오는 방식인 것이죠~
이런형태로 토스트 팝업도 만들 수 있겠습니다!

2. 타입스크립트를 최대한 잘 사용하려고 했던 모습이 보입니다! 굿굿 다만, 올바르게 쓴 것인지는 코드라인 보면서 봐주세요~

3. App 쪽 로직은 아주 간결하고 좋지만.. Page 쪽이 상당히 비즈니스 로직이 꼬여있기도 하고 보기가 별로 좋지 않네요..
custom hooks로 어느정도 덜어낼 수 있을 것으로 보입니다~

4. 나머지는 딱히 리뷰할 부분은 없었습니다. e2e 테스트도 잘 작성했고...
다만 테스트가 조금 적네요..? 다른 시나리오에 있어서도 테스트를 부탁드려용~ 

5. 뭐.. RTK도 이정도면 괜춘쓰~ 다만! 모듈화를 조금 더 신경써주세요~ 한 파일안에서만 말고 도메인 분리!

# 질문 답변

## 첫 번째 질문

```
에러, 로딩처리
각 요청에 대한 에러와, 로딩 상태를 처리하는 부분이 쉽지가 않네요!
각 페이지마다 반복적으로 구현하고 싶지 않아 App에서 에러메시지와, 로딩화면을 보여주려고 시도하였습니다.
에러상태를 에러발생 후 2초 뒤에 초기화 되도록 설정해보았습니다.
좀더 효율적인 에러, 로딩 처리방법이 궁금합니다.
```

1. UX 적으로는 아주 만족스럽습니다 흐흐~ 다만 뒷배경에 opacity 0.3 정도를 가진 background screen fade-in-out 애니메이션이 있어서 부드럽게 보여줬으면 좋겠습니다~

2. 로딩 처리 방법도 괜찮다고 생각이 들구요~ 다만 에러 핸들링 쪽에 조금 문제가 있습니다~ 이 부분은 코드 라인에서 상세히 다뤘습니다~ 확인 부탁쓰!


## 두 번재 질문

```
테스트
로그인, 역 관리 페이지에 대해서만 테스트를 진행하였습니다. 다른 테스트도 구현했어야하는데 시간이 너무 모자르네요..
각 페이지에서 input에 값을 입력하고, submit 버튼을 클릭하였을 때, 삭제버튼을 눌렀을 때 올바른, 올바르지 않은 액션이 dispatch 되는지 까지를 검사하였습니다.
시간이 좀 더 있었다면, 각 페이지에 대하여 통합테스트를 진행한 후 reducer, customhook에 대한 단위테스트를 진행하면 좋겠다고 생각이 들었습니다.
정해진 기간안에 효율적으로 프로젝트를 진행해야 한다고 생각했을 때, 광휘님이 선호하시는 또는 우선순위로 생각하시는 테스트
과정을 공유해주시면 좋겠습니다!
```

촉박한 상황속에서 login, signup 등의 테스트를 하신 것은 아주 칭찬해~ 입니다
form 쪽 부분과 서버쪽에 데이터가 변경되는 부분은 API 유닛테스트 및 시나리오 테스트를 꼭 구현하시는게 좋습니다!

다른 우선순위는 모르겠고, 제일 중요한 부분은 서버에 데이터를 전송하여 변경시키는 기능이 있다면 그 부분은 아주 중요하고 면밀하게 시나리오를 견고하게 작성해서 문제가 일어나지 않는지 체크해야합니다. 왜냐하면 데이터가 변경된다는 것은, 굉장히 민감하거든요. 다르게 수정되면 ㅎㄸ...

## 세 번째 질문

```
커스텀 훅
폼과 인풋이 많다보니 useInput(validator)와 같은식으로 호출하고, 관련 메서드와 값을 반환합니다.
onChange 에서 값을 변경하면서 유효성을 검사하고 그에 대한 에러메시지 상태 값을 hook에서 반환하는데 적절히 이루어졌는지 더 개선할 방법이 있을지도 궁금합니다.
역, 라인 / 구간단위로 커스텀훅을 적용하여 CRUD로직을 분리하였는데 적절한 규모인지 궁금합니다. 좀더 개선할 방법이 있는지도 궁금합니다.🤔

광휘님의 조언이 시급합니다! 😅
```

물론 지금 단위에서도 적절하게 hooks에서 처리하신 걸로보이는데, 조금 더 디벨롭을 하자면 input이 어디서 주로 사용되고, 어떻게 하면 더 쉽게 사용할 수 있을지를 생각해보면 좋겠어용

유효성 검사 로직을 form 단위로 맹글어보는 것은 어떨까요? Context API로 Form을 래핑하고, 하위의 Input에서는 상위 Form Provider가 있을 경우 Validator 로직이 실행되는..? 그러한 상태로요! 저희 비마트의 디자인 시스템에도 이러한 형태가 적용되어있답니다!

그리고, 역 쪽도 코드라인에 피드백을 남겼습니다!

1단계 수고 많으셨어요 브랜님~
앞으로도 홧팅~~

seojihwan committed:
refactor: 모달 Provider를 통해 모달 생성

- Context를 이용하여 useModal, Provider 구현
- 라인 추가, 구간 추가, 컨펌에 대한 모달을 보여주는 과정에서 useModal과 Context API를 이용해 반복되는 로직 제거

seojihwan committed:
refactor: 모달 내부에서 close기능을 갖도록 변경

- 커스텀 훅인 useModal을 Modal내부에서 호출
- export 컨벤션 통일

seojihwan committed:
refactor: 인풋 단위 처리를 폼 단위 처리로 변경

- FormProvider 구현
- useForm 구현
- useInput 제거

seojihwan committed:
refactor: Route 정보를 constants에서 관리하도록 변경

- Routes에서는 props을 통해 route정보를 전달받고, map을 이용하여 중복코드를 제거한다.
- RouteShape과 LinkShape의 타입을 types에서 관리한다

seojihwan committed:
refactor: select단위 처리를 폼 단위 처리로 변경

- SelectField를 구현 (validator X)
- FormProvider에서 Select 태그인경우 유효성 검사를 수행하지 않도록 변경한다
- FormProvider에서 value가 업데이트되기 이전에 validate함수가 동작하는 것을 피하기 위해 validate함수에 value값을 직접 전달
- Select에서 defaultOption을 default value로 지정하기 위해 value값을 -1로 일치시킴

seojihwan committed:
test: render코드의 중복제거

- beforeEach에서 결합된 컴포넌트들을 렌더링
- render의 결과 값이 아닌 screen으로 부터 testing을 위한 메서드들을 가져옴

seojihwan committed:
refactor: 라우트 정보를 Routes 컴포넌트 파일에서 관리

- 라우트 정보를 상수로 관리했을 때 컴포넌트의 생성자 함수가 포함되어 있어 순환참조가 발생

seojihwan committed:
refactor: Station의 타입을 이용하여 중복코드 제거

seojihwan committed:
chore: styled components 바벨 플러그인 추가

- cra overriding 패키지 추가 customize-cra, react-app-rewired

seojihwan committed:
refactor: 인라인 스타일 css prop 으로 변경하기

seojihwan committed:
refactor: enum 대신 object as const 사용

- 재할당 방지
- enum을 사용하는 경우 빌드시 코드가 늘어남
- 역매핑을 사용하지 않음

seojihwan committed:
refactor: api 응답에 대한 타입 구현, 에러 처리 구조 변경

- request함수를 통해 axios요청에 대한 에러처리 및 타입제공
- 파일이름과, 함수 이름의 변경 requestLine => lines
- toolkit에서 request함수에 대한 응답을 분기처리하여 에러 상태를 갱신하도록 수정

seojihwan committed:
refactor: 컨펌창이 버튼 클릭시 닫히도록 수정

seojihwan committed:
refactor: 모든 컬러코드 상수값 이용

seojihwan committed:
refactor: z-index 상수화

seojihwan committed:
refactor: style prop에 대한 타입을 각각 export하도록 변경

seojihwan committed:
refactor: 리덕스 관련 폴더 구조 변경

- action과 리듀서, 타입관련 코드( type, typeguard )들의 코드를 각각 관리

seojihwan committed:
chore: jest의 testMatch 값 설정

seojihwan committed:
test: 인증 action creator 테스트

seojihwan committed:
test: 역 action creator 테스트

seojihwan committed:
test: 노선 action creator 테스트

seojihwan committed:
refactor: Form Provider에서 validator, submit동작 인자로 받도록 변경

- submit에 대한 중복 ( prevent default, 유효성 검사 ) 제거
- validator들을 각 인풋이 관리하지 않고 name을 key로 갖는 객체로 Form Provider에 전달
- FormProvider의 위치를 Form태그 바깥으로 전달
- Form태그는 onSubmit함수를 useForm을 통해 제공 받음
- InputField와 SelectField의 값, onChange함수, 에러메시지를 useForm을 통해 제공받음

seojihwan committed:
refactor: eol 제거

seojihwan commented:
코멘트에 대하여 고민했던 경험과, 궁금한 점을 몇가지 코멘트로 달았습니다!

아래 내용은 리뷰 본문에 대한 답변입니다!

## 테스트 코드

action creator에 대한 단위 테스트를 추가하였습니다! 아직도 부족한 테스트가 많네요 ㅠㅠ😭
9c3603f90815804010428f07c9c9be3f56cc1d66
d5c0f6b27b0fbd5bb0ecc286899eb1a446f54a0e
3b29211c9ff4f63a22ce524ed5c146944ef1bbd4

## Provider를 통해 모달과 폼 구성하기

### 모달

Provider에서 모달을 열고 닫는 메서드를 제공하고 모달을 직접 관리하기 때문에 모달의 상태, 모달 컴포넌트는 숨길 수 있어 코드 구조가 깔끔해졌습니다. 
[코드링크](https://github1s.com/seojihwan/react-subway-map/blob/step1/src/components/ModalProvider/ModalProvider.tsx)

사용하는 곳에서 모달을 열어주면서 모달에 들어갈 children을 넘겨주는 방식을 이용하였습니다~

### 폼

광휘님이 조언해주신 대로 폼 Provider를 만들어 보았습니다.
[코드링크](https://github1s.com/seojihwan/react-subway-map/blob/step1/src/components/FormProvider/FormProvider.tsx)
FormProvider에 각 인풋에 대한 validator들을 전달했을 때 각자의 인풋에 대하여 독립적으로 동작할 수 있도록 구성하였고, 각 Input들의 name값이 key가 되어 value와 errorMessage, validator를 식별하게 됩니다.

비밀번호 확인에 대한 유효성 검사는 Provider 내부에 존재하는 values에 접근해야해서, 고차함수 형태로 validator를 구성하였습니다.

폼 Provider내부의 InputField, SelectField들은 useForm 훅을 이용하여 Provider가 제공하는 value, onChange, ErrorMessage를 활용하였고, Form은 onSubmit을 제공받도록 구성하였습니다.

확실히 form태그에 전달하는 onSubmit함수, 로직, input관련 로직들의 중복이 줄어들었습니다. 
가능하시다면 피드백 부탁드리겠습니다!! ㅎㅎㅎ

우테코 기간중 가장 길게 고민에 빠졌던 미션이었습니다.

광휘님의 리뷰를 반영하면서  객체지향적인 사고력을 좀더 기를 수 있었고 리액트를 통한 구현능력을 키울 수 있었습니다! 감사합니다!!❤

Vallista commented:
(기간이 지낫으나 코맨트 하나만 쓱 던지고 가겠습니다용)

생산성에 대해서 한 번 생각해보세용~
사용자 입장에서 수많은 디자인과 비즈니스를 과연 효과적으로 처리하고 있는 구조인지..

비즈니스는 유연해야합니다. 소프트웨어니까요~

(생각할거리 던지고 탈주)

Vallista commented:
고생 많았어여~

seojihwan commented:
# 리덕스 툴킷
## 내용
- [리덕스 툴킷을 사용한 이유](https://www.notion.so/4-3ee90932e80c4917afe241d38cb36de4)
# Context api
## 내용
- [Context Api를 통한 모달, 폼 프로바이더 구현](https://www.notion.so/4-3ee90932e80c4917afe241d38cb36de4)
